File: test/language/sql_databases/cpp/DbWithRelocationTest.cpp
#include <cstdio>
#include <vector>
#include <string>

#include <sqlite3.h>

#include "gtest/gtest.h"

#include "zserio/StringConvertUtil.h"

#include "sql_databases/db_with_relocation/EuropeDb.h"
#include "sql_databases/db_with_relocation/AmericaDb.h"

namespace sql_databases
{
namespace db_with_relocation
{

class DbWithRelocationTest : public ::testing::Test
{
public:
    DbWithRelocationTest() :
            m_europeDbFileName("db_with_relocation_test_europe.sqlite"),
            m_americaDbFileName("db_with_relocation_test_america.sqlite"),
            m_relocatedSlovakiaTableName("slovakia"),
            m_relocatedCzechiaTableName("czechia"),
            m_europeDb(NULL), m_americaDb(NULL)
    {
        std::remove(m_europeDbFileName.c_str());
        std::remove(m_americaDbFileName.c_str());

        try
        {
            m_europeDb = new EuropeDb(m_europeDbFileName);
            m_europeDb->createSchema();

            AmericaDb::TRelocationMap tableToDbFileNameRelocationMap;
            tableToDbFileNameRelocationMap.insert(std::make_pair(m_relocatedSlovakiaTableName, m_europeDbFileName));
            tableToDbFileNameRelocationMap.insert(std::make_pair(m_relocatedCzechiaTableName, m_europeDbFileName));
            m_americaDb = new AmericaDb(m_americaDbFileName, tableToDbFileNameRelocationMap);
            m_americaDb->createSchema();

            m_attachedDatabasesNames.insert("main");
            m_attachedDatabasesNames.insert("AmericaDb_" + m_relocatedSlovakiaTableName);
            m_attachedDatabasesNames.insert("AmericaDb_" + m_relocatedCzechiaTableName);
        }
        catch (...)
        {
            delete m_europeDb;
            delete m_americaDb;
            throw;
        }
    }

    ~DbWithRelocationTest()
    {
        delete m_europeDb;
        delete m_americaDb;
    }

protected:
    bool isTableInDb(zserio::ISqliteDatabase& database, const std::string& checkTableName)
    {
        sqlite3_stmt* statement;
        std::string sqlQuery = "SELECT name FROM sqlite_master WHERE type='table' AND name='" + checkTableName +
                "'";
        int result = sqlite3_prepare_v2(database.connection(), sqlQuery.c_str(), -1, &statement, NULL);
        if (result != SQLITE_OK)
            return false;

        result = sqlite3_step(statement);
        if (result == SQLITE_DONE || result != SQLITE_ROW)
        {
            sqlite3_finalize(statement);
            return false;
        }

        const unsigned char* readTableName = sqlite3_column_text(statement, 0);
        if (readTableName == NULL || checkTableName.compare(reinterpret_cast<const char*>(readTableName)) != 0)
        {
            sqlite3_finalize(statement);
            return false;
        }

        sqlite3_finalize(statement);

        return true;
    }

    const std::string m_europeDbFileName;
    const std::string m_americaDbFileName;
    const std::string m_relocatedSlovakiaTableName;
    const std::string m_relocatedCzechiaTableName;

    EuropeDb* m_europeDb;
    AmericaDb* m_americaDb;

    std::set<std::string> m_attachedDatabasesNames;
};

TEST_F(DbWithRelocationTest, tableGetters)
{
    CountryMapTable& germanyTable = m_europeDb->getGermany();
    CountryMapTable& usaTable = m_americaDb->getUsa();
    CountryMapTable& canadaTable = m_americaDb->getCanada();
    CountryMapTable& slovakiaTable = m_americaDb->getSlovakia();
    CountryMapTable& czechiaTable = m_americaDb->getCzechia();

    ASSERT_TRUE(&germanyTable != &usaTable);
    ASSERT_TRUE(&usaTable != &canadaTable);
    ASSERT_TRUE(&canadaTable != &slovakiaTable);
    ASSERT_TRUE(&slovakiaTable != &czechiaTable);
}

TEST_F(DbWithRelocationTest, relocatedSlovakiaTable)
{
    ASSERT_FALSE(isTableInDb(*m_americaDb, m_relocatedSlovakiaTableName));
    ASSERT_TRUE(isTableInDb(*m_europeDb, m_relocatedSlovakiaTableName));

    // write to relocated table
    int32_t updateTileId = 1;
    std::vector<CountryMapTableRow> writtenRows(1);
    CountryMapTableRow& row = writtenRows.back();
    row.setTileId(updateTileId);
    Tile writtenTile;
    writtenTile.setVersion('a');
    writtenTile.setData('A');
    row.setTile(writtenTile);
    CountryMapTable& relocatedTable = m_americaDb->getSlovakia();
    relocatedTable.write(writtenRows);

    // update it
    CountryMapTableRow updateRow;
    updateRow.setTileId(updateTileId);
    Tile updatedTile;
    updatedTile.setVersion('b');
    updatedTile.setData('B');
    updateRow.setTile(updatedTile);
    const std::string updateCondition = std::string("tileId=") + zserio::convertToString(updateTileId);
    relocatedTable.update(updateRow, updateCondition);

    // read it back
    std::vector<CountryMapTableRow> readRows;
    relocatedTable.read(readRows);

    ASSERT_EQ(1, readRows.size());
    ASSERT_EQ(updateRow.getTileId(), readRows.front().getTileId());
    ASSERT_EQ(updateRow.getTile(), readRows.front().getTile());
}

TEST_F(DbWithRelocationTest, relocatedCzechiaTable)
{
    ASSERT_FALSE(isTableInDb(*m_americaDb, m_relocatedCzechiaTableName));
    ASSERT_TRUE(isTableInDb(*m_europeDb, m_relocatedCzechiaTableName));

    // write to relocated table
    int32_t updateTileId = 1;
    std::vector<CountryMapTableRow> writtenRows(1);
    CountryMapTableRow& row = writtenRows.back();
    row.setTileId(updateTileId);
    Tile writtenTile;
    writtenTile.setVersion('c');
    writtenTile.setData('C');
    row.setTile(writtenTile);
    CountryMapTable& relocatedTable = m_americaDb->getCzechia();
    relocatedTable.write(writtenRows);

    // update it
    CountryMapTableRow updateRow;
    updateRow.setTileId(updateTileId);
    Tile updatedTile;
    updatedTile.setVersion('d');
    updatedTile.setData('D');
    updateRow.setTile(updatedTile);
    const std::string updateCondition = std::string("tileId=") + zserio::convertToString(updateTileId);
    relocatedTable.update(updateRow, updateCondition);

    // read it back
    std::vector<CountryMapTableRow> readRows;
    relocatedTable.read(readRows);

    ASSERT_EQ(1, readRows.size());
    ASSERT_EQ(updateRow.getTileId(), readRows.front().getTileId());
    ASSERT_EQ(updateRow.getTile(), readRows.front().getTile());
}

TEST_F(DbWithRelocationTest, attachedDatabases)
{
    sqlite3_stmt* stmt = m_americaDb->prepareStatement("PRAGMA database_list");

    while (sqlite3_step(stmt) == SQLITE_ROW)
    {
        const char* databaseName = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 1));
        ASSERT_EQ(1, m_attachedDatabasesNames.count(databaseName));
        m_attachedDatabasesNames.erase(databaseName);
    }

    sqlite3_finalize(stmt);

    ASSERT_EQ(1, m_attachedDatabasesNames.size());
    ASSERT_EQ(0, m_attachedDatabasesNames.count("main"));
}

} // namespace db_with_relocation
} // namespace sql_databases

File: test/language/sql_databases/python/DbWithRelocationTest.py
import unittest
import os

from testutils import getZserioApi, getApiDir

class DbWithRelocationTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "sql_databases.zs").db_with_relocation
        cls._europeDbFileName = os.path.join(getApiDir(os.path.dirname(__file__)),
                                             "db_with_relocation_test_europe.sqlite")
        cls._americaDbFileName = os.path.join(getApiDir(os.path.dirname(__file__)),
                                              "db_with_relocation_test_america.sqlite")

    def setUp(self):
        if os.path.exists(self._europeDbFileName):
            os.remove(self._europeDbFileName)
        if os.path.exists(self._americaDbFileName):
            os.remove(self._americaDbFileName)

        self._europeDb = self.api.EuropeDb.from_file(self._europeDbFileName)
        self._europeDb.create_schema()

        tableToDbFileNameRelocationMap = {self.RELOCATED_SLOVAKIA_TABLE_NAME : self._europeDbFileName,
                                          self.RELOCATED_CZECHIA_TABLE_NAME : self._europeDbFileName}
        self._americaDb = self.api.AmericaDb.from_file(self._americaDbFileName, tableToDbFileNameRelocationMap)
        self._americaDb.create_schema()

    def tearDown(self):
        self._americaDb.close()
        self._europeDb.close()

    def testTableGetters(self):
        germanyTable = self._europeDb.germany
        self.assertNotEqual(None, germanyTable)

        usaTable = self._americaDb.usa
        self.assertNotEqual(None, usaTable)

        canadaTable = self._americaDb.canada
        self.assertNotEqual(None, canadaTable)

        slovakiaTable = self._americaDb.slovakia
        self.assertNotEqual(None, slovakiaTable)

        czechiaTable = self._americaDb.czechia
        self.assertNotEqual(None, czechiaTable)

    def testRelocatedSlovakiaTable(self):
        # check that americaDb does not contain relocated table
        self.assertFalse(self._isTableInDb(self._americaDb, self.RELOCATED_SLOVAKIA_TABLE_NAME))

        # check that europeDb does contain relocated table
        self.assertTrue(self._isTableInDb(self._europeDb, self.RELOCATED_SLOVAKIA_TABLE_NAME))

        # write to relocated table
        updateTileId = 1
        writtenRows = [(updateTileId, self.api.Tile(ord('a'), ord('A')))]
        relocatedTable = self._americaDb.slovakia
        relocatedTable.write(writtenRows)

        # update it
        updatedRows = [(updateTileId, self.api.Tile(ord('b'), ord('B')))]
        updateCondition = "tileId=" + str(updateTileId)
        relocatedTable.update(updatedRows[0], updateCondition)

        # read it back
        readRows = relocatedTable.read()
        numReadRows = 0
        for readRow in readRows:
            self.assertEqual(updatedRows[numReadRows], readRow)
            numReadRows += 1
        self.assertTrue(len(updatedRows), numReadRows)

    def testRelocatedCzechiaTable(self):
        # check that americaDb does not contain relocated table
        self.assertFalse(self._isTableInDb(self._americaDb, self.RELOCATED_CZECHIA_TABLE_NAME))

        # check that europeDb does contain relocated table
        self.assertTrue(self._isTableInDb(self._europeDb, self.RELOCATED_CZECHIA_TABLE_NAME))

        # write to relocated table
        updateTileId = 1
        writtenRows = [(updateTileId, self.api.Tile(ord('c'), ord('C')))]
        relocatedTable = self._americaDb.czechia
        relocatedTable.write(writtenRows)

        # update it
        updatedRows = [(updateTileId, self.api.Tile(ord('d'), ord('D')))]
        updateCondition = "tileId=" + str(updateTileId)
        relocatedTable.update(updatedRows[0], updateCondition)

        # read it back
        readRows = relocatedTable.read()
        numReadRows = 0
        for readRow in readRows:
            self.assertEqual(updatedRows[numReadRows], readRow)
            numReadRows += 1
        self.assertTrue(len(updatedRows), numReadRows)

    def testAttachedDatabases(self):
        attachedDatabaseNames = ["main",
                                 "AmericaDb_" + self.RELOCATED_SLOVAKIA_TABLE_NAME,
                                 "AmericaDb_" + self.RELOCATED_CZECHIA_TABLE_NAME]
        sqlQuery = "PRAGMA database_list"
        rows = self._americaDb.connection.cursor().execute(sqlQuery)
        for row in rows:
            self.assertIn(row[1], attachedDatabaseNames)
            attachedDatabaseNames.remove(row[1])
        self.assertEqual(1, len(attachedDatabaseNames))
        self.assertNotIn("main", attachedDatabaseNames)

    @staticmethod
    def _isTableInDb(database, tableName):
        # check if database does contain table
        sqlQuery = "SELECT name FROM sqlite_master WHERE type='table' AND name='" + tableName + "'"
        rows = database.connection.cursor().execute(sqlQuery)
        for row in rows:
            if len(row) == 1 and row[0] == tableName:
                return True

        return False

    RELOCATED_SLOVAKIA_TABLE_NAME = "slovakia"
    RELOCATED_CZECHIA_TABLE_NAME = "czechia"
